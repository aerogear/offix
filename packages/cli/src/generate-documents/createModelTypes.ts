import endent from "endent";

const getModelProperties = (schema: any) => {
    const fieldMap = schema.properties;
    const keys = Object.keys(fieldMap);

    return keys
      .map(fieldName => {
        const s = fieldMap[fieldName];
        const name = s.isRequired ? fieldName : `${fieldName}?`;
        return `${name}: ${s.type}`;
      })
      .join(";\n");
};

export const createModelType = (schema: any) => {
    const { name: modelName, primaryKey } = schema;

    return endent`
      export interface ${modelName} {
        ${getModelProperties(schema)}
      }

      export type ${modelName}Create = ${primaryKey ? `Omit<${modelName}, "${primaryKey}">` : modelName};
      export type ${modelName}Change =  ${primaryKey ? `Pick<${modelName}, "${primaryKey}"> & ` : ""}Partial<${modelName}Create>;

    `;
};
