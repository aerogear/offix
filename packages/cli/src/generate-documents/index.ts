export * from "./createJsonSchema";
export * from "./createModelTypes";
