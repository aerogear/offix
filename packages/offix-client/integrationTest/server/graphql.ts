import express from "express";
import { VoyagerServer } from "@aerogear/voyager-server";
import http from "http";
import { SubscriptionServer } from "subscriptions-transport-ws";
import { execute, subscribe } from "graphql";

import { typeDefs, resolvers } from "./schema";

const PORT = 4000;

export function startServer() {
  const app = express();

  const apolloServer = VoyagerServer({ typeDefs, resolvers });
  const httpServer = http.createServer(app);
  apolloServer.applyMiddleware({ app });

  return new Promise(resolve => {
    httpServer.listen({ port: PORT }, async () => {
      const subscriptionServer = new SubscriptionServer({
        execute,
        subscribe,
        schema: apolloServer.schema
      }, {
        server: httpServer,
        path: "/graphql"
      });
      resolve(httpServer);
    });
  });
}
