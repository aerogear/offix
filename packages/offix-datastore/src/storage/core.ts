import { v1 as uuidv1 } from "uuid";

export function generateId() {
    return uuidv1();
}
