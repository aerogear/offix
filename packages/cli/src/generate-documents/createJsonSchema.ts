import { ModelDefinition, getPrimaryKey } from "@graphback/core";
import {
  isNonNullType,
  GraphQLOutputType,
  getNullableType
} from "graphql";
import { convertToTsType } from "../utils";

const getFieldParameters = (fieldName: string, type: GraphQLOutputType): any => {
  const options: any = {};

  options.key = fieldName;

  if (isNonNullType(type)) {
    type = getNullableType(type);
    options.isRequired = true;
  }

  return { type: convertToTsType(type), ...options };
};

const getModelProperties = (model: ModelDefinition, primaryKey: string) => {
  const fieldMap = model.graphqlType.getFields();
  const keys = Object.keys(fieldMap);
  // get a list of relationships
  const relNames = model.relationships.map(r => r.ownerField.name);

  // loop through graphback relationship types
  const relationships = model.relationships
    // filter out oneToMany relationships
    .filter(r => r.kind !== "oneToMany")
    // generate key/value pair with foreign key as key
    .map(r => {
      const fieldOptions = getFieldParameters(
        r.relationForeignKey!,
        r.relationType
      );
      fieldOptions.relationship = r.relationType;
      return { [r.relationForeignKey!]: fieldOptions };
    });

  const generatedProperties = keys
    // filter out autogenerated relationship properties
    .filter(fieldName => !relNames.includes(fieldName))
    .map(fieldName => {
      const fieldOptions = getFieldParameters(
        fieldName,
        fieldMap[fieldName].type
       );
      if (fieldName === primaryKey) {
        fieldOptions.primary = true;
      }
      return { [fieldName]: fieldOptions };
    })
    // add previously created relationship key/values
    .concat(relationships)
    .reduce((prev, current) => ({ ...prev, ...current }), {});

  generatedProperties._version = {
    type: "string",
    key: "_version",
    isRequired: true
  };

  generatedProperties._lastUpdatedAt = {
    type: "number",
    key: "_lastUpdatedAt",
    isRequired: true
  };

  return generatedProperties;
};

export const createJsonSchema = (model: ModelDefinition) => {
  const primaryKey = getPrimaryKey(model.graphqlType).name;

  return {
    name: model.graphqlType.name,
    version: 1,
    type: "object",
    primaryKey: primaryKey,
    properties: getModelProperties(model, primaryKey)
  };
};
