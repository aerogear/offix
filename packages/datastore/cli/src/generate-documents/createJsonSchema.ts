import { ModelDefinition, getPrimaryKey } from "@graphback/core";
import {
  isNonNullType,
  GraphQLOutputType,
  getNullableType
} from "graphql";
import { convertToTsType } from "../utils";

const getFieldParameters = (fieldName: string, type: GraphQLOutputType ): any => {
  const options: any = {};

  options.key = fieldName;

  if (isNonNullType(type)) {
    type = getNullableType(type);
    options.isRequired = true;
  }

  return { type: convertToTsType(type), ...options };
};

const getModelProperties = (model: ModelDefinition, primaryKey: string) => {
  const fieldMap = model.graphqlType.getFields();
  const keys = Object.keys(fieldMap);

  const relationships = model.relationships
    .filter(r => !keys.includes(r.ownerField.name))
    .map(r => {
      const fieldOptions = getFieldParameters(
        r.relationForeignKey!,
        r.relationType
      );
      fieldOptions.relationship = r.relationType;
      return { [r.relationForeignKey!]: fieldOptions };
    });
  const relNames = model.relationships.map(r => r.ownerField.name);

  const generatedProperties = keys
    .filter(fieldName => !relNames.includes(fieldName))
    .map(fieldName => {
      const fieldOptions = getFieldParameters(
        fieldName,
        fieldMap[fieldName].type
       );
      if (fieldName === primaryKey) {
        fieldOptions.primary = true;
      }
      return { [fieldName]: fieldOptions };
    })
    .concat(relationships)
    .reduce((prev, current) => ({ ...prev, ...current }), {});

  generatedProperties._version = {
    type: "string",
    key: "_version",
    isRequired: true
  };
  return generatedProperties;
};

// TODO refactor use GraphQL type instead of model
// definition
export const createJsonSchema = (model: ModelDefinition) => {
  const primaryKey = getPrimaryKey(model.graphqlType).name;

  return {
    name: model.graphqlType.name,
    version: 1,
    type: "object",
    primaryKey: primaryKey,
    properties: getModelProperties(model, primaryKey)
  };
};
