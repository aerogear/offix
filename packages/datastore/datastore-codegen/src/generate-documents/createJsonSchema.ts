import { ModelDefinition } from "@graphback/core";
import {
    isNonNullType,
    GraphQLOutputType,
    getNullableType
} from "graphql";
import { convertToTsType } from "../utils";


const getFieldParameters = (fieldName: string, type: GraphQLOutputType): any => {
    const options: any = {};

    // TODO handle relationships

    options.key = fieldName;

    if (isNonNullType(type)) {
        type = getNullableType(type);
        options.isRequired = true;
    }
    if (type.toString() === "ID") {
        options.index = true;
        options.primary = true;
    }

    return { type: convertToTsType(type), ...options };
};

const getModelProperties = (model: ModelDefinition) => {
    const fieldMap = model.graphqlType.getFields();

    return Object.keys(fieldMap)
        .map(fieldName => ({
            [fieldName]: getFieldParameters(fieldName, fieldMap[fieldName].type)
        }))
        .reduce((prev, current) => ({ ...prev, ...current }), {});
};

export const createJsonSchema = (model: ModelDefinition) => {
    return {
        name: model.graphqlType.name,
        version: 1,
        type: "object",
        primaryKey: "id",
        properties: getModelProperties(model)
    };
};
