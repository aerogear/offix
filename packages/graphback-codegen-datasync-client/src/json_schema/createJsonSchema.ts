import { ModelDefinition } from "@graphback/core";

export const createJsonSchema = (model: ModelDefinition) => {
    return {
        name: model.graphqlType.name,
        properties: getModelProperties(model)
    }
}

const getModelProperties = (model: ModelDefinition) => {
    const fieldMap = model.graphqlType.getFields();
    return Object.keys(fieldMap)
        .map(fieldName => ({ [fieldName]: { test: "test" } }))
        .reduce((prev, current) => ({ ...prev, ...current }), {});
}
