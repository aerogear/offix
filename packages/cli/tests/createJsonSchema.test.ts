import { GraphbackCoreMetadata, ModelDefinition } from "@graphback/core";
import { buildSchema } from "graphql";
import { createJsonSchema } from "../src/generate-documents/createJsonSchema";

const noteFragment = (annotation: string, array: boolean = true) => (`
  """
  @model
  @datasync
  """
  type Note {
    _id: GraphbackObjectID!
    title: String!
    description: String!
    ${annotation}
    ${array ? "comments: [Comment]!" : "comment: Comment!"}
  }
`);

const commentFragment = (annotation: string) => (`
  """
  @model
  @datasync
  """
  type Comment {
    _id: GraphbackObjectID!
    title: String!
    description: String!
    ${annotation}
    noteComment: Note!
  }
`);

test("createJsonSchema (@oneToOne - default)", () => {
  const modelSDL = `
    scalar GraphbackObjectID
    ${noteFragment(`
      """
      @oneToOne()
      """
    `, false)}
    ${commentFragment("")}
  `;

  const metadata = new GraphbackCoreMetadata({ crudMethods: {} }, buildSchema(modelSDL));
  const models = metadata.getModelDefinitions();
  const note = models.find((m: ModelDefinition) => m.graphqlType.name === "Note");
  const comment = models.find((m: ModelDefinition) => m.graphqlType.name === "Comment");

  const noteJsonSchema: any = createJsonSchema(note as ModelDefinition);
  expect(noteJsonSchema.properties.commentId).toBeDefined();
  const commentJsonSchema: any = createJsonSchema(comment as ModelDefinition);
  expect(commentJsonSchema.properties.noteId).not.toBeDefined();
});

test("createJsonSchema (@oneToOne - key specified)", () => {
  const modelSDL = `
    scalar GraphbackObjectID
    ${noteFragment(`
      """
      @oneToOne(key: 'commentId')
      """
    `, false)}
    ${commentFragment("")}
  `;

  const metadata = new GraphbackCoreMetadata({ crudMethods: {} }, buildSchema(modelSDL));
  const models = metadata.getModelDefinitions();
  const note = models.find((m: ModelDefinition) => m.graphqlType.name === "Note");
  const comment = models.find((m: ModelDefinition) => m.graphqlType.name === "Comment");

  const noteJsonSchema: any = createJsonSchema(note as ModelDefinition);
  expect(noteJsonSchema.properties.commentId).toBeDefined();
  expect(noteJsonSchema).toMatchSnapshot();

  const commentJsonSchema: any = createJsonSchema(comment as ModelDefinition);
  expect(commentJsonSchema.properties.noteId).not.toBeDefined();
  expect(commentJsonSchema).toMatchSnapshot();
});

test("createJsonSchema (@oneToMany - default key)", () => {
  const modelSDL = `
    scalar GraphbackObjectID
    ${noteFragment(`
      """
      @oneToMany(field: 'noteComment')
      """
    `)}
    ${commentFragment("")}
  `;

  const metadata = new GraphbackCoreMetadata({ crudMethods: {} }, buildSchema(modelSDL));
  const models = metadata.getModelDefinitions();
  const note = models.find((m: ModelDefinition) => m.graphqlType.name === "Note");
  const comment = models.find((m: ModelDefinition) => m.graphqlType.name === "Comment");

  const noteJsonSchema: any = createJsonSchema(note as ModelDefinition);
  expect(noteJsonSchema.properties.comment).not.toBeDefined();
  expect(noteJsonSchema).toMatchSnapshot();

  const commentJsonSchema: any = createJsonSchema(comment as ModelDefinition);
  expect(commentJsonSchema.properties.noteComment).not.toBeDefined();
  expect(commentJsonSchema.properties.noteCommentId).toBeDefined();
  expect(commentJsonSchema).toMatchSnapshot();
});

test("createJsonSchema (@oneToMany - key specified)", () => {
  const modelSDL = `
    scalar GraphbackObjectID
    ${noteFragment(`
      """
      @oneToMany(field: 'noteComment', key: 'noteId')
      """
    `)}
    ${commentFragment("")}
  `;

  const metadata = new GraphbackCoreMetadata({ crudMethods: {} }, buildSchema(modelSDL));
  const models = metadata.getModelDefinitions();
  const note = models.find((m: ModelDefinition) => m.graphqlType.name === "Note");
  const comment = models.find((m: ModelDefinition) => m.graphqlType.name === "Comment");

  const noteJsonSchema: any = createJsonSchema(note as ModelDefinition);
  expect(noteJsonSchema.properties.comment).not.toBeDefined();
  expect(noteJsonSchema).toMatchSnapshot();

  const commentJsonSchema: any = createJsonSchema(comment as ModelDefinition);
  expect(commentJsonSchema.properties.noteId).toBeDefined();
  expect(commentJsonSchema).toMatchSnapshot();
});

test("createJsonSchema (@oneToMany & @manyToOne - default key)", () => {
  const modelSDL = `
    scalar GraphbackObjectID
    ${noteFragment(`
      """
      @oneToMany(field: 'noteComment')
      """
    `)}
    ${commentFragment(`
      """
      @manyToOne(field: 'comments')
      """
    `)}
  `;

  const metadata = new GraphbackCoreMetadata({ crudMethods: {} }, buildSchema(modelSDL));
  const models = metadata.getModelDefinitions();
  const note = models.find((m: ModelDefinition) => m.graphqlType.name === "Note");
  const comment = models.find((m: ModelDefinition) => m.graphqlType.name === "Comment");

  const noteJsonSchema: any = createJsonSchema(note as ModelDefinition);
  expect(noteJsonSchema.properties.comment).not.toBeDefined();
  expect(noteJsonSchema).toMatchSnapshot();

  const commentJsonSchema: any = createJsonSchema(comment as ModelDefinition);
  expect(commentJsonSchema.properties.noteComment).not.toBeDefined();
  expect(commentJsonSchema.properties.noteCommentId).toBeDefined();
  expect(commentJsonSchema).toMatchSnapshot();
});

test("createJsonSchema (@oneToMany & @manyToOne - key specified)", () => {
  const modelSDL = `
    scalar GraphbackObjectID
    ${noteFragment(`
      """
      @oneToMany(field: 'noteComment', key: 'noteId')
      """
    `)}
    ${commentFragment(`
      """
      @manyToOne(field: 'comments', key: 'noteId')
      """
    `)}
  `;

  const metadata = new GraphbackCoreMetadata({ crudMethods: {} }, buildSchema(modelSDL));
  const models = metadata.getModelDefinitions();
  const note = models.find((m: ModelDefinition) => m.graphqlType.name === "Note");
  const comment = models.find((m: ModelDefinition) => m.graphqlType.name === "Comment");

  const noteJsonSchema: any = createJsonSchema(note as ModelDefinition);
  expect(noteJsonSchema.properties.comment).not.toBeDefined();
  expect(noteJsonSchema).toMatchSnapshot();

  const commentJsonSchema: any = createJsonSchema(comment as ModelDefinition);
  expect(commentJsonSchema.properties.noteId).toBeDefined();
  expect(commentJsonSchema).toMatchSnapshot();
});
