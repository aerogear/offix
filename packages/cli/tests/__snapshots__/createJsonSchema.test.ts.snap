// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createJsonSchema (@oneToMany & @manyToOne - default key) 1`] = `
Object {
  "name": "Note",
  "primaryKey": "_id",
  "properties": Object {
    "_id": Object {
      "isRequired": true,
      "key": "_id",
      "primary": true,
      "type": "string",
    },
    "_lastUpdatedAt": Object {
      "isRequired": true,
      "key": "_lastUpdatedAt",
      "type": "number",
    },
    "_version": Object {
      "isRequired": true,
      "key": "_version",
      "type": "string",
    },
    "description": Object {
      "isRequired": true,
      "key": "description",
      "type": "string",
    },
    "title": Object {
      "isRequired": true,
      "key": "title",
      "type": "string",
    },
  },
  "type": "object",
  "version": 1,
}
`;

exports[`createJsonSchema (@oneToMany & @manyToOne - default key) 2`] = `
Object {
  "name": "Comment",
  "primaryKey": "_id",
  "properties": Object {
    "_id": Object {
      "isRequired": true,
      "key": "_id",
      "primary": true,
      "type": "string",
    },
    "_lastUpdatedAt": Object {
      "isRequired": true,
      "key": "_lastUpdatedAt",
      "type": "number",
    },
    "_version": Object {
      "isRequired": true,
      "key": "_version",
      "type": "string",
    },
    "description": Object {
      "isRequired": true,
      "key": "description",
      "type": "string",
    },
    "noteCommentId": Object {
      "key": "noteCommentId",
      "relationship": "Note",
      "type": "string",
    },
    "title": Object {
      "isRequired": true,
      "key": "title",
      "type": "string",
    },
  },
  "type": "object",
  "version": 1,
}
`;

exports[`createJsonSchema (@oneToMany & @manyToOne - key specified) 1`] = `
Object {
  "name": "Note",
  "primaryKey": "_id",
  "properties": Object {
    "_id": Object {
      "isRequired": true,
      "key": "_id",
      "primary": true,
      "type": "string",
    },
    "_lastUpdatedAt": Object {
      "isRequired": true,
      "key": "_lastUpdatedAt",
      "type": "number",
    },
    "_version": Object {
      "isRequired": true,
      "key": "_version",
      "type": "string",
    },
    "description": Object {
      "isRequired": true,
      "key": "description",
      "type": "string",
    },
    "title": Object {
      "isRequired": true,
      "key": "title",
      "type": "string",
    },
  },
  "type": "object",
  "version": 1,
}
`;

exports[`createJsonSchema (@oneToMany & @manyToOne - key specified) 2`] = `
Object {
  "name": "Comment",
  "primaryKey": "_id",
  "properties": Object {
    "_id": Object {
      "isRequired": true,
      "key": "_id",
      "primary": true,
      "type": "string",
    },
    "_lastUpdatedAt": Object {
      "isRequired": true,
      "key": "_lastUpdatedAt",
      "type": "number",
    },
    "_version": Object {
      "isRequired": true,
      "key": "_version",
      "type": "string",
    },
    "description": Object {
      "isRequired": true,
      "key": "description",
      "type": "string",
    },
    "noteId": Object {
      "key": "noteId",
      "relationship": "Note",
      "type": "string",
    },
    "title": Object {
      "isRequired": true,
      "key": "title",
      "type": "string",
    },
  },
  "type": "object",
  "version": 1,
}
`;

exports[`createJsonSchema (@oneToMany - default key) 1`] = `
Object {
  "name": "Note",
  "primaryKey": "_id",
  "properties": Object {
    "_id": Object {
      "isRequired": true,
      "key": "_id",
      "primary": true,
      "type": "string",
    },
    "_lastUpdatedAt": Object {
      "isRequired": true,
      "key": "_lastUpdatedAt",
      "type": "number",
    },
    "_version": Object {
      "isRequired": true,
      "key": "_version",
      "type": "string",
    },
    "description": Object {
      "isRequired": true,
      "key": "description",
      "type": "string",
    },
    "title": Object {
      "isRequired": true,
      "key": "title",
      "type": "string",
    },
  },
  "type": "object",
  "version": 1,
}
`;

exports[`createJsonSchema (@oneToMany - default key) 2`] = `
Object {
  "name": "Comment",
  "primaryKey": "_id",
  "properties": Object {
    "_id": Object {
      "isRequired": true,
      "key": "_id",
      "primary": true,
      "type": "string",
    },
    "_lastUpdatedAt": Object {
      "isRequired": true,
      "key": "_lastUpdatedAt",
      "type": "number",
    },
    "_version": Object {
      "isRequired": true,
      "key": "_version",
      "type": "string",
    },
    "description": Object {
      "isRequired": true,
      "key": "description",
      "type": "string",
    },
    "noteCommentId": Object {
      "key": "noteCommentId",
      "relationship": "Note",
      "type": "string",
    },
    "title": Object {
      "isRequired": true,
      "key": "title",
      "type": "string",
    },
  },
  "type": "object",
  "version": 1,
}
`;

exports[`createJsonSchema (@oneToMany - key specified) 1`] = `
Object {
  "name": "Note",
  "primaryKey": "_id",
  "properties": Object {
    "_id": Object {
      "isRequired": true,
      "key": "_id",
      "primary": true,
      "type": "string",
    },
    "_lastUpdatedAt": Object {
      "isRequired": true,
      "key": "_lastUpdatedAt",
      "type": "number",
    },
    "_version": Object {
      "isRequired": true,
      "key": "_version",
      "type": "string",
    },
    "description": Object {
      "isRequired": true,
      "key": "description",
      "type": "string",
    },
    "title": Object {
      "isRequired": true,
      "key": "title",
      "type": "string",
    },
  },
  "type": "object",
  "version": 1,
}
`;

exports[`createJsonSchema (@oneToMany - key specified) 2`] = `
Object {
  "name": "Comment",
  "primaryKey": "_id",
  "properties": Object {
    "_id": Object {
      "isRequired": true,
      "key": "_id",
      "primary": true,
      "type": "string",
    },
    "_lastUpdatedAt": Object {
      "isRequired": true,
      "key": "_lastUpdatedAt",
      "type": "number",
    },
    "_version": Object {
      "isRequired": true,
      "key": "_version",
      "type": "string",
    },
    "description": Object {
      "isRequired": true,
      "key": "description",
      "type": "string",
    },
    "noteId": Object {
      "key": "noteId",
      "relationship": "Note",
      "type": "string",
    },
    "title": Object {
      "isRequired": true,
      "key": "title",
      "type": "string",
    },
  },
  "type": "object",
  "version": 1,
}
`;

exports[`createJsonSchema (@oneToOne - key specified) 1`] = `
Object {
  "name": "Note",
  "primaryKey": "_id",
  "properties": Object {
    "_id": Object {
      "isRequired": true,
      "key": "_id",
      "primary": true,
      "type": "string",
    },
    "_lastUpdatedAt": Object {
      "isRequired": true,
      "key": "_lastUpdatedAt",
      "type": "number",
    },
    "_version": Object {
      "isRequired": true,
      "key": "_version",
      "type": "string",
    },
    "commentId": Object {
      "key": "commentId",
      "relationship": "Comment",
      "type": "string",
    },
    "description": Object {
      "isRequired": true,
      "key": "description",
      "type": "string",
    },
    "title": Object {
      "isRequired": true,
      "key": "title",
      "type": "string",
    },
  },
  "type": "object",
  "version": 1,
}
`;

exports[`createJsonSchema (@oneToOne - key specified) 2`] = `
Object {
  "name": "Comment",
  "primaryKey": "_id",
  "properties": Object {
    "_id": Object {
      "isRequired": true,
      "key": "_id",
      "primary": true,
      "type": "string",
    },
    "_lastUpdatedAt": Object {
      "isRequired": true,
      "key": "_lastUpdatedAt",
      "type": "number",
    },
    "_version": Object {
      "isRequired": true,
      "key": "_version",
      "type": "string",
    },
    "description": Object {
      "isRequired": true,
      "key": "description",
      "type": "string",
    },
    "noteComment": Object {
      "isRequired": true,
      "key": "noteComment",
      "type": "string",
    },
    "title": Object {
      "isRequired": true,
      "key": "title",
      "type": "string",
    },
  },
  "type": "object",
  "version": 1,
}
`;
